# Task 3: Data Cleaning

## Retail Store Sales Dataset

### Objective
The objective of this project is to transform a deliberately messy retail sales dataset into a clean, consistent, and analysis-ready dataset using Python, Pandas, and NumPy.

### Data Cleaning Process
1. Load and inspect the dataset
2. Generate a data quality report
3. Handle missing values
4. Remove duplicate records
5. Standardize inconsistent values
6. Correct data types
7. Detect and handle outliers
8. Validate the cleaned dataset
9. Compare data quality before and after cleaning
10. Export the cleaned dataset

In [83]:
import pandas as pd
import numpy as np

# Display all columns
pd.set_option("display.max_columns", None)

# Load dataset
df = pd.read_csv("../data/retail_store_sales.csv")

print("Dataset loaded successfully.")
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Dataset loaded successfully.
Rows: 12575
Columns: 11


In [84]:
# Display first 5 rows
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False


In [85]:
# Dataset information
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  str    
 1   Customer ID       12575 non-null  str    
 2   Category          12575 non-null  str    
 3   Item              11362 non-null  str    
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  str    
 8   Location          12575 non-null  str    
 9   Transaction Date  12575 non-null  str    
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(1), str(7)
memory usage: 1.1+ MB


In [86]:
# Basic statistical summary
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Transaction ID,12575,12575,TXN_6867343,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Customer ID,12575,25,CUST_05,544,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Category,12575,8,Furniture,1591,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Item,11362,200,Item_2_BEV,126,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Price Per Unit,11966.0,NaN,NaN,NaN,23.365912,10.743519,5.0,14.0,23.0,33.5,41.0
Quantity,11971.0,NaN,NaN,NaN,5.53638,2.857883,1.0,3.0,6.0,8.0,10.0
Total Spent,11971.0,NaN,NaN,NaN,129.652577,94.750697,5.0,51.0,108.5,192.0,410.0
Payment Method,12575,3,Cash,4310,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Location,12575,2,Online,6354,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Transaction Date,12575,1114,2022-05-30,26,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [87]:
# Check column names
print("Column names:")
for column in df.columns:
    print("-", column)

Column names:
- Transaction ID
- Customer ID
- Category
- Item
- Price Per Unit
- Quantity
- Total Spent
- Payment Method
- Location
- Transaction Date
- Discount Applied


In [88]:
# Check missing values
missing_values = df.isnull().sum()

print("Missing values per column:")
print(missing_values)

Missing values per column:
Transaction ID         0
Customer ID            0
Category               0
Item                1213
Price Per Unit       609
Quantity             604
Total Spent          604
Payment Method         0
Location               0
Transaction Date       0
Discount Applied    4199
dtype: int64


In [89]:
# Check duplicate rows
duplicate_count = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_count}")

Duplicate rows: 0


In [90]:
# Check unique values in categorical/object columns
for column in df.select_dtypes(include="object").columns:
    print(f"\n--- {column} ---")
    print(df[column].value_counts(dropna=False).head(20))


--- Transaction ID ---
Transaction ID
TXN_6867343    1
TXN_3731986    1
TXN_9303719    1
TXN_9458126    1
TXN_4575373    1
TXN_7482416    1
TXN_3652209    1
TXN_1372952    1
TXN_9728486    1
TXN_2722661    1
TXN_8776416    1
TXN_5422631    1
TXN_5874772    1
TXN_4413070    1
TXN_2490363    1
TXN_1809665    1
TXN_7563311    1
TXN_9634894    1
TXN_4396807    1
TXN_4206593    1
Name: count, dtype: int64

--- Customer ID ---
Customer ID
CUST_05    544
CUST_24    543
CUST_13    534
CUST_08    533
CUST_09    519
CUST_15    519
CUST_16    515
CUST_23    513
CUST_01    507
CUST_18    507
CUST_20    507
CUST_11    503
CUST_22    501
CUST_10    501
CUST_21    498
CUST_12    498
CUST_07    491
CUST_02    488
CUST_17    487
CUST_19    487
Name: count, dtype: int64

--- Category ---
Category
Furniture                             1591
Electric household essentials         1591
Food                                  1588
Milk Products                         1584
Butchers                             

C:\Users\kishw\AppData\Local\Temp\ipykernel_12792\2684459376.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for column in df.select_dtypes(include="object").columns:


In [91]:
for column in ["Category", "Payment Method", "Location", "Discount Applied"]:
    print(f"\n{'='*50}")
    print(column)
    print(f"{'='*50}")
    print(df[column].value_counts(dropna=False))


Category
Category
Furniture                             1591
Electric household essentials         1591
Food                                  1588
Milk Products                         1584
Butchers                              1568
Beverages                             1567
Computers and electric accessories    1558
Patisserie                            1528
Name: count, dtype: int64

Payment Method
Payment Method
Cash              4310
Digital Wallet    4144
Credit Card       4121
Name: count, dtype: int64

Location
Location
Online      6354
In-store    6221
Name: count, dtype: int64

Discount Applied
Discount Applied
True     4219
NaN      4199
False    4157
Name: count, dtype: int64


In [92]:
numeric_columns = [
    "Price Per Unit",
    "Quantity",
    "Total Spent"
]

print(df[numeric_columns].describe().T)

                  count        mean        std  min   25%    50%    75%    max
Price Per Unit  11966.0   23.365912  10.743519  5.0  14.0   23.0   33.5   41.0
Quantity        11971.0    5.536380   2.857883  1.0   3.0    6.0    8.0   10.0
Total Spent     11971.0  129.652577  94.750697  5.0  51.0  108.5  192.0  410.0


In [93]:
for column in numeric_columns:
    print(f"\n{column}")
    print("Negative values:", (df[column] < 0).sum())
    print("Zero values:", (df[column] == 0).sum())
    print("Minimum:", df[column].min())
    print("Maximum:", df[column].max())


Price Per Unit
Negative values: 0
Zero values: 0
Minimum: 5.0
Maximum: 41.0

Quantity
Negative values: 0
Zero values: 0
Minimum: 1.0
Maximum: 10.0

Total Spent
Negative values: 0
Zero values: 0
Minimum: 5.0
Maximum: 410.0


In [94]:
calculated_total = df["Price Per Unit"] * df["Quantity"]

comparison = pd.DataFrame({
    "Price Per Unit": df["Price Per Unit"],
    "Quantity": df["Quantity"],
    "Total Spent": df["Total Spent"],
    "Calculated Total": calculated_total
})

comparison.head(20)

,Price Per Unit,Quantity,Total Spent,Calculated Total
0,18.5,10.0,185.0,185.0
1,29.0,9.0,261.0,261.0
2,21.5,2.0,43.0,43.0
3,27.5,9.0,247.5,247.5
4,12.5,7.0,87.5,87.5
5,NaN,10.0,200.0,NaN
6,5.0,8.0,40.0,40.0
7,33.5,NaN,NaN,NaN
8,27.5,1.0,27.5,27.5
9,36.5,3.0,109.5,109.5


In [95]:
print(df["Transaction Date"].head(20))
print("\nData type:", df["Transaction Date"].dtype)
print("Unique dates:", df["Transaction Date"].nunique())

0     2024-04-08
1     2023-07-23
2     2022-10-05
3     2022-05-07
4     2022-10-02
5     2023-11-30
6     2023-06-10
7     2024-04-02
8     2023-04-26
9     2024-03-14
10    2024-12-14
11    2025-01-12
12    2023-09-09
13    2022-05-20
14    2022-05-22
15    2022-05-11
16    2024-11-16
17    2022-04-17
18    2022-02-07
19    2025-01-13
Name: Transaction Date, dtype: str

Data type: str
Unique dates: 1114


In [96]:
numeric_columns = [
    "Price Per Unit",
    "Quantity",
    "Total Spent"
]

outlier_report = []

for column in numeric_columns:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ][column].count()

    outlier_report.append({
        "Column": column,
        "Q1": Q1,
        "Q3": Q3,
        "IQR": IQR,
        "Lower Bound": lower_bound,
        "Upper Bound": upper_bound,
        "Outlier Count": outliers
    })

outlier_report = pd.DataFrame(outlier_report)

outlier_report

,Column,Q1,Q3,IQR,Lower Bound,Upper Bound,Outlier Count
0,Price Per Unit,14.0,33.5,19.5,-15.25,62.75,0
1,Quantity,3.0,8.0,5.0,-4.50,15.50,0
2,Total Spent,51.0,192.0,141.0,-160.50,403.50,60


In [97]:
total_spent_outliers = df[
    (df["Total Spent"] < -160.50) |
    (df["Total Spent"] > 403.50)
]

total_spent_outliers[
    ["Category", "Item", "Price Per Unit", "Quantity", "Total Spent"]
].sort_values("Total Spent", ascending=False).head(20)

,Category,Item,Price Per Unit,Quantity,Total Spent
27,Furniture,Item_25_FUR,41.0,10.0,410.0
133,Furniture,Item_25_FUR,41.0,10.0,410.0
339,Food,Item_25_FOOD,41.0,10.0,410.0
869,Beverages,Item_25_BEV,41.0,10.0,410.0
1060,Furniture,Item_25_FUR,41.0,10.0,410.0
1088,Beverages,Item_25_BEV,41.0,10.0,410.0
1468,Butchers,Item_25_BUT,41.0,10.0,410.0
1505,Electric household essentials,Item_25_EHE,41.0,10.0,410.0
1568,Butchers,Item_25_BUT,41.0,10.0,410.0
1950,Furniture,Item_25_FUR,41.0,10.0,410.0


In [98]:
print("Number of Total Spent outliers:", len(total_spent_outliers))
print("Outlier Total Spent values:")
print(total_spent_outliers["Total Spent"].value_counts().sort_index())

Number of Total Spent outliers: 60
Outlier Total Spent values:
Total Spent
410.0    60
Name: count, dtype: int64


## 1. Data Quality Report

Before cleaning the dataset, a data quality assessment was performed to identify missing values, duplicate records, incorrect data types, and numerical anomalies.

The initial inspection identified:
- Missing values in `Item`, `Price Per Unit`, `Quantity`, `Total Spent`, and `Discount Applied`.
- No duplicate rows.
- `Transaction Date` stored as a string instead of a datetime type.
- `Discount Applied` stored as an object despite representing Boolean values.
- No negative or zero values in the numerical columns.
- 60 IQR-based outliers in `Total Spent`, all with a value of 410. These were investigated and determined to be valid transactions because 410 can result from the maximum observed price of 41 multiplied by the maximum quantity of 10.

The findings from this assessment were used to determine the appropriate cleaning strategy for each column.

In [99]:
# Generate initial data quality report

quality_report = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str).values,
    "Missing Values": df.isnull().sum().values,
    "Missing %": (df.isnull().sum() / len(df) * 100).round(2).values,
    "Unique Values": df.nunique(dropna=True).values
})

quality_report

,Column,Data Type,Missing Values,Missing %,Unique Values
0,Transaction ID,str,0,0.00,12575
1,Customer ID,str,0,0.00,25
2,Category,str,0,0.00,8
3,Item,str,1213,9.65,200
4,Price Per Unit,float64,609,4.84,25
5,Quantity,float64,604,4.80,10
6,Total Spent,float64,604,4.80,227
7,Payment Method,str,0,0.00,3
8,Location,str,0,0.00,2
9,Transaction Date,str,0,0.00,1114


In [100]:
# Store initial dataset statistics for before-vs-after comparison

before_rows = len(df)
before_duplicates = df.duplicated().sum()
before_nulls = df.isnull().sum().sum()
before_dtypes = df.dtypes.astype(str).to_dict()

print("Before Cleaning")
print("----------------")
print("Rows:", before_rows)
print("Duplicate Rows:", before_duplicates)
print("Total Missing Values:", before_nulls)

Before Cleaning
----------------
Rows: 12575
Duplicate Rows: 0
Total Missing Values: 7229


## 2. Missing Data Handling

Different strategies are used for missing values depending on the data type and business meaning of each column.

- **Item:** Missing item names will be filled with `"Unknown"` because the exact product cannot be reliably inferred from the available information.
- **Price Per Unit:** Missing numerical values will be replaced using the median. The median is less sensitive to extreme values than the mean and is appropriate for transactional price data.
- **Quantity:** Missing quantities will be replaced using the median because quantity is numerical and the median provides a robust estimate without being strongly affected by unusual transactions.
- **Total Spent:** Where `Price Per Unit` and `Quantity` are available, `Total Spent` will be calculated as `Price Per Unit × Quantity`. Any remaining missing values will be filled using the median.
- **Discount Applied:** Missing Boolean values will be filled using the mode because the column represents a categorical True/False indicator and the most frequent observed value provides a simple consistent replacement.

No rows will be deleted because the missing values can be handled without unnecessarily reducing the dataset size.

In [101]:
df["Item"] = df["Item"].fillna("Unknown")

print("Missing Item values:", df["Item"].isnull().sum())

Missing Item values: 0


In [102]:
price_median = df["Price Per Unit"].median()

df["Price Per Unit"] = df["Price Per Unit"].fillna(price_median)

print("Price Per Unit median used:", price_median)
print("Remaining missing values:", df["Price Per Unit"].isnull().sum())

Price Per Unit median used: 23.0
Remaining missing values: 0


In [103]:
quantity_median = df["Quantity"].median()

df["Quantity"] = df["Quantity"].fillna(quantity_median)

print("Quantity median used:", quantity_median)
print("Remaining missing values:", df["Quantity"].isnull().sum())

Quantity median used: 6.0
Remaining missing values: 0


In [104]:
missing_total_before = df["Total Spent"].isnull().sum()

df["Total Spent"] = df["Price Per Unit"] * df["Quantity"]

print("Missing Total Spent before recalculation:", missing_total_before)
print("Missing Total Spent after recalculation:", df["Total Spent"].isnull().sum())

Missing Total Spent before recalculation: 604
Missing Total Spent after recalculation: 0


In [105]:
discount_mode = df["Discount Applied"].mode()[0]

print("Discount Applied mode:", discount_mode)

Discount Applied mode: True


In [106]:
df["Discount Applied"] = df["Discount Applied"].fillna(discount_mode)

print("Remaining missing values:", df["Discount Applied"].isnull().sum())

Remaining missing values: 0


In [107]:
print("Missing values after treatment:")
print(df.isnull().sum())

Missing values after treatment:
Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64


## 3. Duplicate Removal

Duplicate records were checked using all columns in the dataset.

The initial inspection identified 0 duplicate rows. Therefore, no records were removed during this stage.

Keeping the original row count ensures that valid transactions are not unnecessarily discarded.

In [108]:
# Check and remove duplicate rows

rows_before_duplicates = len(df)
duplicates_before = df.duplicated().sum()

df = df.drop_duplicates()

rows_after_duplicates = len(df)
duplicates_removed = rows_before_duplicates - rows_after_duplicates

print("Duplicate rows before cleaning:", duplicates_before)
print("Duplicate rows removed:", duplicates_removed)
print("Duplicate rows after cleaning:", df.duplicated().sum())

Duplicate rows before cleaning: 0
Duplicate rows removed: 0
Duplicate rows after cleaning: 0


## 4. Data Standardisation

Categorical values were inspected for inconsistent capitalization, spelling, and formatting.

The dataset already uses consistent category labels such as `Cash`, `Credit Card`, `Digital Wallet`, `Online`, and `In-store`. Therefore, no category values require manual mapping.

As a preventive cleaning step, leading and trailing whitespace will be removed from text columns.

In [109]:
# Standardize only actual text columns

text_columns = [
    "Transaction ID",
    "Customer ID",
    "Category",
    "Item",
    "Payment Method",
    "Location"
]

for column in text_columns:
    df[column] = df[column].str.strip()

print("Text standardisation completed successfully.")


Text standardisation completed successfully.


In [110]:
print("Data types after correction:")
print(df.dtypes)

Data types after correction:
Transaction ID          str
Customer ID             str
Category                str
Item                    str
Price Per Unit      float64
Quantity            float64
Total Spent         float64
Payment Method          str
Location                str
Transaction Date        str
Discount Applied     object
dtype: object


In [81]:
print("\nDataset shape:")
print(df.shape)

print("\nRemaining missing values:")
print(df.isnull().sum())


Dataset shape:
(12575, 11)

Remaining missing values:
Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64


## 5. Data Type Correction

Data types are corrected according to the meaning of each column.

- `Transaction ID` and `Customer ID` are kept as strings because they are identifiers.
- `Transaction Date` is converted from string to datetime.
- `Quantity` is converted from float to integer because transaction quantities are whole numbers.
- `Discount Applied` is converted from object to Boolean because it represents whether a discount was applied.
- `Price Per Unit` and `Total Spent` remain floating-point numerical values.

In [112]:
# Correct data types

df["Quantity"] = df["Quantity"].astype("int64")

df["Transaction Date"] = pd.to_datetime(
    df["Transaction Date"],
    errors="coerce"
)

df["Discount Applied"] = df["Discount Applied"].astype("bool")

print("Data types corrected successfully.")

Data types corrected successfully.


In [113]:
# Verify corrected data types

print("Data types after correction:")
print(df.dtypes)

Data types after correction:
Transaction ID                 str
Customer ID                    str
Category                       str
Item                           str
Price Per Unit             float64
Quantity                     int64
Total Spent                float64
Payment Method                 str
Location                       str
Transaction Date    datetime64[us]
Discount Applied              bool
dtype: object


## 6. Outlier Detection and Handling

The Interquartile Range (IQR) method was used to identify potential outliers in the numerical columns.

The analysis found:
- `Price Per Unit`: 0 outliers
- `Quantity`: 0 outliers
- `Total Spent`: 60 outliers

All 60 `Total Spent` outliers had a value of 410. Although the IQR method classified these observations as statistical outliers, they represent valid transactions because the dataset contains a maximum price of 41 and a maximum quantity of 10, making 410 a valid possible transaction total.

Therefore, these observations were retained rather than removed or capped. This prevents the loss of legitimate high-value transactions.

In [34]:
# IQR-based outlier detection

outlier_summary = []

for column in ["Price Per Unit", "Quantity", "Total Spent"]:
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = df[
        (df[column] < lower_bound) |
        (df[column] > upper_bound)
    ]

    outlier_summary.append({
        "Column": column,
        "Lower Bound": lower_bound,
        "Upper Bound": upper_bound,
        "Outlier Count": len(outliers)
    })

outlier_summary = pd.DataFrame(outlier_summary)

outlier_summary

,Column,Lower Bound,Upper Bound,Outlier Count
0,Price Per Unit,-13.0,59.0,0
1,Quantity,-4.5,15.5,0
2,Total Spent,-150.5,397.5,56


In [35]:
# Verify retained Total Spent outliers

total_spent_outliers = df[df["Total Spent"] > 403.5]

print("Retained Total Spent outliers:", len(total_spent_outliers))
print("Unique outlier values:", total_spent_outliers["Total Spent"].unique())

Retained Total Spent outliers: 56
Unique outlier values: [410.]


In [36]:
# Final data quality validation

print("Dataset shape:", df.shape)
print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:", df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

Dataset shape: (12575, 11)

Missing values:
Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64

Duplicate rows: 0

Data types:
Transaction ID          str
Customer ID             str
Category                str
Item                    str
Price Per Unit      float64
Quantity            float64
Total Spent         float64
Payment Method          str
Location                str
Transaction Date        str
Discount Applied     object
dtype: object


In [37]:
# Check numerical columns for invalid values

print("Negative values:")
print((df[["Price Per Unit", "Quantity", "Total Spent"]] < 0).sum())

print("\nZero values:")
print((df[["Price Per Unit", "Quantity", "Total Spent"]] == 0).sum())

Negative values:
Price Per Unit    0
Quantity          0
Total Spent       0
dtype: int64

Zero values:
Price Per Unit    0
Quantity          0
Total Spent       0
dtype: int64


## 7. Final Data Validation

After completing the cleaning process, the dataset was validated to ensure that missing values, duplicate records, invalid numerical values, and data type issues were resolved.

The cleaned dataset should contain the same number of valid transaction records as the original dataset because no rows needed to be deleted.

In [114]:
# Final validation

print("Dataset shape:", df.shape)

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:", df.duplicated().sum())

print("\nData types:")
print(df.dtypes)

Dataset shape: (12575, 11)

Missing values:
Transaction ID      0
Customer ID         0
Category            0
Item                0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
dtype: int64

Duplicate rows: 0

Data types:
Transaction ID                 str
Customer ID                    str
Category                       str
Item                           str
Price Per Unit             float64
Quantity                     int64
Total Spent                float64
Payment Method                 str
Location                       str
Transaction Date    datetime64[us]
Discount Applied              bool
dtype: object


In [115]:
# Check for invalid numerical values

print("Negative values:")
print((df[["Price Per Unit", "Quantity", "Total Spent"]] < 0).sum())

print("\nZero values:")
print((df[["Price Per Unit", "Quantity", "Total Spent"]] == 0).sum())

Negative values:
Price Per Unit    0
Quantity          0
Total Spent       0
dtype: int64

Zero values:
Price Per Unit    0
Quantity          0
Total Spent       0
dtype: int64


## 8. Before vs After Cleaning Summary

The following comparison summarizes the improvement in data quality after cleaning.

The cleaning process removed missing values, preserved all valid transaction records, confirmed that no duplicate rows were present, and corrected the data types of the relevant columns.

In [116]:
# Final Before vs After summary

comparison_summary = pd.DataFrame({
    "Metric": [
        "Row Count",
        "Total Missing Values",
        "Duplicate Rows",
        "Correct Data Types"
    ],
    "Before Cleaning": [
        before_rows,
        before_nulls,
        duplicates_before,
        "8/11"
    ],
    "After Cleaning": [
        len(df),
        df.isnull().sum().sum(),
        df.duplicated().sum(),
        "11/11"
    ]
})

comparison_summary

,Metric,Before Cleaning,After Cleaning
0,Row Count,12575,12575
1,Total Missing Values,7229,0
2,Duplicate Rows,0,0
3,Correct Data Types,8/11,11/11


## 9. Export Cleaned Dataset

The cleaned dataset is exported as a new CSV file so that the original raw dataset remains unchanged.

The resulting file is ready for downstream analysis and visualization.

In [118]:
# Save cleaned dataset

output_path = "../data/cleaned_dataset.csv"

df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved successfully to: {output_path}")

Cleaned dataset saved successfully to: ../data/cleaned_dataset.csv


In [119]:
# Verify exported dataset

cleaned_df = pd.read_csv(output_path)

print("Exported dataset shape:", cleaned_df.shape)
print("\nMissing values in exported dataset:")
print(cleaned_df.isnull().sum().sum())

Exported dataset shape: (12575, 11)

Missing values in exported dataset:
0
